python/paddle/distributed/auto_parallel/sharding.py

```diff
@@ -349,17 +373,82 @@ def apply_gradients(self, params_grads):
             self._inner_opt._grad_clip.has_dist_param = has_dist_param
             self._inner_opt._grad_clip.has_not_dist_param = has_not_dist_param
         self._inner_opt.apply_gradients(new_params_grads)
+
         for (
+            slice_param_list,
             shard_param,
             fused_param,
         ) in all_gather_param_info_list:
-            allgather_value = paddle._C_ops.all_gather(
-                shard_param, self._sharding_group.id, self._sharding_degree
-            )
+            if self._strategy.sharding.enable_overlap:
+                opt_op = slice_param_list[-1].all_used_ops()[-1]
+                # NOTE: add dependency between opt op and allgather_value for correctness
+                tmp = paddle._C_ops.nop(opt_op.results()[0])
+                tmp.get_defining_op().set_execution_stream(
+                    AutoParallelStreamType.SHARDING_STREAM.value
+                )
+
+                allgather_value = paddle._C_ops.all_gather(
+                    shard_param, self._sharding_group.id, self._sharding_degree
+                )
+                allgather_value.get_defining_op().set_execution_stream(
+                    AutoParallelStreamType.SHARDING_STREAM.value
+                )
```

Comment:
shard_param and slice_param are share_var, the dependency should be constructed automatically.